### PR TITLE
mon: remove reports if we get the msg of the osd

### PR DESCRIPTION
**Pull request by simon-rock**
fix : https://tracker.ceph.com/issues/47380
       https://tracker.ceph.com/issues/24531
Signed-off-by: Simon Gao <simon29rock@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -2640,6 +2640,25 @@ bool OSDMonitor::preprocess_query(MonOpRequestRef op)
...
+	  mon->no_reply(report_op);
+	}
+      }
+      ceph_assert(fi.reporters.empty());
+      dout(10) << " removing last failure_info for osd." << sid
+	       << dendl;
+      failure_info.erase(sid);
+    }
+  }
+
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc:2661]: i failed to find any analysis or reasonings in the commit message or in the tracker ticket. why do you think the peer osds cannot withdraw the failure report from monitor after the failed one is reachable again?

Comment (simon-rock) [reply src/mon/OSDMonitor.cc:2661]: @tchaikov  I find the issue and a path to fix it.
I think this problem still exists. 
You can see this patch for some information.
https://github.com/ceph/ceph/pull/28177

I think we need to clean up the records of the osd if  we have received its message.

**Comment by jecluis** [comment]
I'll review all this pr and reasoning this week.

**Comment by neha-ojha** [comment]
@jecluis @tchaikov ping

**Comment by jdurgin** [comment]
@tchaikov ping

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (tchaikov) [reply src/mon/OSDMonitor.cc:2661]: @simon-rock i tried to contact you via your mail address in the Signed-off-by line, but no replies so far. my question still holds.

Comment (tchaikov) [reply src/mon/OSDMonitor.cc:2661]: i think we should not rely on the message from the target osd to cancel the stale reports. what if the target osd is relatively quiet from the leader's perspective after being reported unresponsive?

i am creating #40033 as an alternative to address this issue. to check the `failed_for` time, if it takes way to long to mark an osd down, we just give up and drop all failure reports.

**Comment by tchaikov** [comment]
could fill the commit message with more details, like

> failure_info keeps strong references of the MOSDFailure messages
> sent by osd or peon monitors, whenever monitor starts to handle
> an MOSDFailure message, it registers it in its OpTracker. and
> the failure report messageis unregistered when monitor acks them
> by either canceling them or replying the reporters with a new
> osdmap marking the target osd down. but if this does not happen,
> the failure reports just pile up in OpTracker. and monitor considers
> them as slow ops. and they are reported as SLOW_OPS health warning.
> 
> in theory, it does not take long to mark an unresponsive osd down if
> we have enough reporters. but there is chance, that a reporter fails
> to cancel its report before it reboots, and the monitor also fails
> to collect enough reports and mark the target osd down. so the
> target osd never gets an osdmap marking it down, so it won't send
> an alive message to monitor to fix this.

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (tchaikov) [reply src/mon/OSDMonitor.cc:2661]: discussed with @simon-rock offline. this approach is better. but couple nits

- better off cancel the failure report in `preprocess_beacon()`. as an active osd is supposed to send beacon messages periodically.
- no need to reply no_op
- no need to lookup in `failure_info` twice.

**Comment by gregsfortytwo** [comment]
No, this doesn't work. OSDs heartbeat their peers so that we can detect network partitions *between* OSDs. An OSD which is not available to its peers needs to be marked down, even if it can route to the monitor.

**Comment by tchaikov** [comment]
> An OSD which is not available to its peers needs to be marked down, even if it can route to the monitor.

if an OSD is not marked done in the period of a beacon, it's safe to drop the failure info.

> Even if that weren't the case, this only fixes things if the OSD in question happens to be connected to the lead monitor, which is a 1/3 or 1/5 chance.

beacons are always routed to leader.

**Comment by tchaikov** [comment]
> No, this doesn't work. OSDs heartbeat their peers so that we can detect network partitions *between* OSDs. An OSD which is not available to its peers needs to be marked down, even if it can route to the monitor.

ahh, right. makes sense to me!

**Comment by tchaikov** [comment]
closing in favor of #40033

**Pull request closed without merging**